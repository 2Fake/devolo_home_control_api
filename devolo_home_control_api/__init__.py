"""the devolo Home Control API."""
from importlib.metadata import PackageNotFoundError, version

try:
    __version__ = version("devolo_home_control_api")
except PackageNotFoundError:
    # package is not installed - e.g. pulled and run locally
    __version__ = "0.0.0"
