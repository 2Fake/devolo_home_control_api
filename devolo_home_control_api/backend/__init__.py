MESSAGE_TYPES = {
    "devolo.BinarySensor": "_binary_sensor",
    "devolo.BinarySwitch": "_binary_switch",
    "devolo.Blinds": "_multi_level_switch",
    "devolo.DevicesPage": "_inspect_devices",
    "devolo.DewpointSensor": "_multi_level_sensor",
    "devolo.Dimmer": "_multi_level_switch",
    "devolo.Grouping": "_grouping",
    "devolo.HumidityBarValue": "_humidity_bar",
    "devolo.HumidityBarZone": "_humidity_bar",
    "devolo.mprm.gw.GatewayAccessibilityFI": "_gateway_accessible",
    "devolo.Meter": "_meter",
    "devolo.MildewSensor": "_binary_sensor",
    "devolo.MultiLevelSensor": "_multi_level_sensor",
    "devolo.MultiLevelSwitch": "_multi_level_switch",
    "devolo.RemoteControl": "_remote_control",
    "devolo.SirenMultiLevelSwitch": "_multi_level_switch",
    "devolo.ShutterMovementFI": "_binary_sensor",
    "devolo.ValveTemperatureSensor": "_multi_level_sensor",
    "devolo.VoltageMultiLevelSensor": "_multi_level_sensor",
    "devolo.WarningBinaryFI": "_binary_sensor",
    "hdm": "_device_state",
    "acs.hdm": "_automatic_calibration",
    "bas.hdm": "_binary_async",
    "bss.hdm": "_binary_sync",
    "cps.hdm": "_parameter",
    "gds.hdm": "_general_device",
    "lis.hdm": "_led",
    "mas.hdm": "_multilevel_async",
    "mss.hdm": "_multilevel_sync",
    "ps.hdm": "_protection",
    "stmss.hdm": "_multilevel_sync",
    "sts.hdm": "_switch_type",
    "trs.hdm": "_temperature_report",
    "vfs.hdm": "_led"
}
