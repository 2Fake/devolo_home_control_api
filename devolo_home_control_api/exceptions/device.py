class MprmDeviceCommunicationError(Exception):
    """ Communicating to a device via mPRM failed. """


class MprmDeviceNotFoundError(Exception):
    """ A device like this was not found. """


class WrongElementError(Exception):
    """ This element was not meant for this property. """


class SwitchingProtected(Exception):
    """ This device is protected against remote switching. """
