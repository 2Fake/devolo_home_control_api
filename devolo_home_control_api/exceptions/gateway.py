class GatewayOfflineError(Exception):
    """ The requested gateway is offline and cannot be used. """
