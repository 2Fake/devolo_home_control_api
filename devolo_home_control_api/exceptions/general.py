class WrongCredentialsError(Exception):
    """ Wrong credentials were used. """


class WrongUrlError(Exception):
    """ Wrong URL was used. """
