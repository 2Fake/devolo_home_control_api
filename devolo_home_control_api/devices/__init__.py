"""Devices appearing in a devolo Home Control setup."""
from .gateway import Gateway
from .zwave import Zwave

__all__ = ["Gateway", "Zwave"]
