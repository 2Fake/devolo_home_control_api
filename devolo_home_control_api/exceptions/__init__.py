"""Expections used in this package."""
from .device import SwitchingProtected, WrongElementError
from .gateway import GatewayOfflineError
from .general import WrongCredentialsError, WrongUrlError

__all__ = ["GatewayOfflineError", "SwitchingProtected", "WrongCredentialsError", "WrongElementError", "WrongUrlError"]
