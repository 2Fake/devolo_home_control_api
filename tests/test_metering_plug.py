"""Test interacting with a metering plug."""
import json
import sys
from datetime import datetime, timezone

import pytest
from dateutil import tz
from requests_mock import Mocker
from syrupy.assertion import SnapshotAssertion

from devolo_home_control_api.exceptions import SwitchingProtected
from devolo_home_control_api.homecontrol import HomeControl

from . import HOMECONTROL_URL, Subscriber, load_fixture
from .mocks import WEBSOCKET

ELEMENT_ID = "hdm:ZWave:CBC56091/2"
FIXTURE = load_fixture("homecontrol_binary_switch")
TIMEZONE = tz.gettz()


@pytest.mark.skipif(sys.version_info < (3, 8), reason="Tests with snapshots need at least Python 3.8")
def test_getting_devices(local_gateway: HomeControl, snapshot: SnapshotAssertion) -> None:
    """Test getting binary switch devices."""
    assert local_gateway.binary_switch_devices == snapshot


def test_switching(local_gateway: HomeControl, gateway_ip: str, requests_mock: Mocker) -> None:
    """Test switching a binary switch."""
    response = FIXTURE["success"]
    requests_mock.post(f"http://{gateway_ip}/remote/json-rpc", json=response)
    subscriber = Subscriber(ELEMENT_ID)
    local_gateway.publisher.register(ELEMENT_ID, subscriber)
    metering_plug = local_gateway.devices[ELEMENT_ID]

    binary_switch = metering_plug.binary_switch_property[f"devolo.BinarySwitch:{ELEMENT_ID}"]
    state = binary_switch.state
    last_activity = binary_switch.last_activity
    assert binary_switch.set(state=not state)
    assert state != binary_switch.state
    assert last_activity != binary_switch.last_activity

    state = binary_switch.state
    last_activity = binary_switch.last_activity
    WEBSOCKET.recv_packet(json.dumps(FIXTURE["switch_event"]))
    assert state != binary_switch.state
    assert last_activity != binary_switch.last_activity
    assert subscriber.update.call_count == 1


def test_switching_same_state(local_gateway: HomeControl, gateway_ip: str, requests_mock: Mocker) -> None:
    """Test switching a binary switch to the same state."""
    response = FIXTURE["fail"]
    requests_mock.post(f"http://{gateway_ip}/remote/json-rpc", json=response)
    subscriber = Subscriber(ELEMENT_ID)
    local_gateway.publisher.register(ELEMENT_ID, subscriber)

    metering_plug = local_gateway.devices[ELEMENT_ID]
    binary_switch = metering_plug.binary_switch_property[f"devolo.BinarySwitch:{ELEMENT_ID}"]
    state = binary_switch.state
    assert not binary_switch.set(state=not state)
    assert state == binary_switch.state


def test_switching_protected(local_gateway: HomeControl) -> None:
    """Test switching a binary switch that is protected."""
    metering_plug = local_gateway.devices[ELEMENT_ID]
    binary_switch = metering_plug.binary_switch_property[f"devolo.BinarySwitch:{ELEMENT_ID}"]
    binary_switch.enabled = False
    with pytest.raises(SwitchingProtected):
        binary_switch.set(state=True)


def test_consumption(local_gateway: HomeControl) -> None:
    """Test consumption a mitering plug."""
    element = next(
        item
        for item in load_fixture("homecontrol_device_details")["result"]["items"]
        if item["UID"] == f"devolo.Meter:{ELEMENT_ID}"
    )
    subscriber = Subscriber(ELEMENT_ID)
    local_gateway.publisher.register(ELEMENT_ID, subscriber)

    metering_plug = local_gateway.devices[ELEMENT_ID]
    consumption = metering_plug.consumption_property[f"devolo.Meter:{ELEMENT_ID}"]
    assert consumption.current == element["properties"]["currentValue"]
    assert consumption.total == element["properties"]["totalValue"]
    assert consumption.total_since == datetime.fromtimestamp(
        element["properties"]["sinceTime"] / 1000, tz=timezone.utc
    ).replace(tzinfo=tz.gettz("Europe/Berlin"))

    last_activity = consumption.last_activity
    WEBSOCKET.recv_packet(json.dumps(FIXTURE["current_event"]))
    assert consumption.current == FIXTURE["current_event"]["properties"]["property.value.new"]
    assert last_activity != consumption.last_activity
    assert subscriber.update.call_count == 1

    last_activity = consumption.last_activity
    WEBSOCKET.recv_packet(json.dumps(FIXTURE["total_event"]))
    assert consumption.total == FIXTURE["total_event"]["properties"]["property.value.new"]
    assert last_activity != consumption.last_activity
    assert subscriber.update.call_count == 2

    WEBSOCKET.recv_packet(json.dumps(FIXTURE["total_since"]))
    assert consumption.total_since == datetime.fromtimestamp(
        FIXTURE["total_since"]["properties"]["property.value.new"] / 1000, tz=timezone.utc
    ).replace(tzinfo=tz.gettz("Europe/Berlin"))
    assert subscriber.update.call_count == 3


def test_voltage(local_gateway: HomeControl) -> None:
    """Test voltage sensor of a metering plug."""
    element = next(
        item
        for item in load_fixture("homecontrol_device_details")["result"]["items"]
        if item["UID"] == f"devolo.VoltageMultiLevelSensor:{ELEMENT_ID}"
    )
    subscriber = Subscriber(ELEMENT_ID)
    local_gateway.publisher.register(ELEMENT_ID, subscriber)

    metering_plug = local_gateway.devices[ELEMENT_ID]
    voltage = metering_plug.multi_level_sensor_property[f"devolo.VoltageMultiLevelSensor:{ELEMENT_ID}"]
    assert voltage.value == element["properties"]["value"]

    last_activity = voltage.last_activity
    WEBSOCKET.recv_packet(json.dumps(FIXTURE["voltage_event"]))
    assert voltage.value == FIXTURE["voltage_event"]["properties"]["property.value.new"]
    assert last_activity != voltage.last_activity


def test_changing_general_device_settings(local_gateway: HomeControl) -> None:
    """Test changing general device settings."""
    metering_plug = local_gateway.devices[ELEMENT_ID]
    events_enabled = metering_plug.settings_property["general_device_settings"].events_enabled
    icon = metering_plug.settings_property["general_device_settings"].icon
    name = metering_plug.settings_property["general_device_settings"].name
    zone_id = metering_plug.settings_property["general_device_settings"].zone_id
    WEBSOCKET.recv_packet(json.dumps(FIXTURE["general_device_settings"]))
    assert events_enabled != metering_plug.settings_property["general_device_settings"].events_enabled
    assert icon != metering_plug.settings_property["general_device_settings"].icon
    assert name != metering_plug.settings_property["general_device_settings"].name
    assert zone_id != metering_plug.settings_property["general_device_settings"].zone_id


def test_setting_general_device_settings(remote_gateway: HomeControl, requests_mock: Mocker) -> None:
    """Test setting general device settings."""
    response = FIXTURE["success"]
    requests_mock.post(f"{HOMECONTROL_URL}/remote/json-rpc", json=response)
    metering_plug = remote_gateway.devices[ELEMENT_ID]
    events_enabled = metering_plug.settings_property["general_device_settings"].events_enabled
    icon = metering_plug.settings_property["general_device_settings"].icon
    name = metering_plug.settings_property["general_device_settings"].name
    zone_id = metering_plug.settings_property["general_device_settings"].zone_id
    assert metering_plug.settings_property["general_device_settings"].set(icon="icon_15")
    assert metering_plug.settings_property["general_device_settings"].events_enabled == events_enabled
    assert icon != metering_plug.settings_property["general_device_settings"].icon
    assert name == metering_plug.settings_property["general_device_settings"].name
    assert zone_id == metering_plug.settings_property["general_device_settings"].zone_id


def test_setting_general_device_settings_same_state(remote_gateway: HomeControl, requests_mock: Mocker) -> None:
    """Test setting general device settings to the same state."""
    response = FIXTURE["fail"]
    requests_mock.post(f"{HOMECONTROL_URL}/remote/json-rpc", json=response)
    metering_plug = remote_gateway.devices[ELEMENT_ID]
    icon = metering_plug.settings_property["general_device_settings"].icon
    assert not metering_plug.settings_property["general_device_settings"].set(icon="icon_15")
    assert icon == metering_plug.settings_property["general_device_settings"].icon


def test_changing_led_settings(local_gateway: HomeControl) -> None:
    """Test changing led settings."""
    metering_plug = local_gateway.devices[ELEMENT_ID]
    WEBSOCKET.recv_packet(json.dumps(FIXTURE["led"]))
    assert not metering_plug.settings_property["led"].led_setting


def test_setting_led(remote_gateway: HomeControl, requests_mock: Mocker) -> None:
    """Test setting led settings."""
    response = FIXTURE["success"]
    requests_mock.post(f"{HOMECONTROL_URL}/remote/json-rpc", json=response)
    metering_plug = remote_gateway.devices[ELEMENT_ID]
    led_setting = metering_plug.settings_property["led"].led_setting
    assert metering_plug.settings_property["led"].set(led_setting=False)
    assert metering_plug.settings_property["led"].led_setting != led_setting


def test_setting_led_same_state(remote_gateway: HomeControl, requests_mock: Mocker) -> None:
    """Test setting led settings to the same state."""
    response = FIXTURE["fail"]
    requests_mock.post(f"{HOMECONTROL_URL}/remote/json-rpc", json=response)
    metering_plug = remote_gateway.devices[ELEMENT_ID]
    led_setting = metering_plug.settings_property["led"].led_setting
    assert not metering_plug.settings_property["led"].set(led_setting=True)
    assert metering_plug.settings_property["led"].led_setting == led_setting


def test_changing_protection_mode(local_gateway: HomeControl) -> None:
    """Test changing the protection mode."""
    metering_plug = local_gateway.devices[ELEMENT_ID]
    binary_switch = metering_plug.binary_switch_property[f"devolo.BinarySwitch:{ELEMENT_ID}"]
    WEBSOCKET.recv_packet(json.dumps(FIXTURE["protection"]))
    WEBSOCKET.recv_packet(json.dumps(FIXTURE["gui_enabled"]))
    assert not metering_plug.settings_property["protection"].remote_switching
    assert not binary_switch.enabled

    FIXTURE["protection"]["properties"]["property.name"] = "localSwitch"
    FIXTURE["protection"]["properties"]["com.prosyst.mbs.services.remote.event.sequence.number"] = 2
    metering_plug = local_gateway.devices[ELEMENT_ID]
    WEBSOCKET.recv_packet(json.dumps(FIXTURE["protection"]))
    assert not metering_plug.settings_property["protection"].local_switching


def test_protection_mode(remote_gateway: HomeControl, requests_mock: Mocker) -> None:
    """Test setting the protection mode."""
    response = FIXTURE["success"]
    requests_mock.post(f"{HOMECONTROL_URL}/remote/json-rpc", json=response)
    metering_plug = remote_gateway.devices[ELEMENT_ID]
    local_switching = metering_plug.settings_property["protection"].local_switching
    remote_switching = metering_plug.settings_property["protection"].remote_switching
    assert metering_plug.settings_property["protection"].set(local_switching=False)
    assert metering_plug.settings_property["protection"].local_switching != local_switching
    assert metering_plug.settings_property["protection"].remote_switching == remote_switching


def test_protection_mode_same_state(remote_gateway: HomeControl, requests_mock: Mocker) -> None:
    """Test setting the protection mode to the same state."""
    response = FIXTURE["fail"]
    requests_mock.post(f"{HOMECONTROL_URL}/remote/json-rpc", json=response)
    metering_plug = remote_gateway.devices[ELEMENT_ID]
    local_switching = metering_plug.settings_property["protection"].local_switching
    assert not metering_plug.settings_property["protection"].set(local_switching=False)
    assert metering_plug.settings_property["protection"].local_switching == local_switching


def test_changing_expert_settings(local_gateway: HomeControl) -> None:
    """Test changing expert settings."""
    metering_plug = local_gateway.devices[ELEMENT_ID]
    param_changed = metering_plug.settings_property["param_changed"].param_changed
    WEBSOCKET.recv_packet(json.dumps(FIXTURE["param_changed"]))
    assert param_changed != metering_plug.settings_property["param_changed"].param_changed


def test_changing_flash_mode_settings(local_gateway: HomeControl) -> None:
    """Test changing flash mode settings."""
    metering_plug = local_gateway.devices[ELEMENT_ID]
    WEBSOCKET.recv_packet(json.dumps(FIXTURE["flash_mode"]))
    assert metering_plug.settings_property["flash_mode"].value == 2


def test_getting_properties(local_gateway: HomeControl) -> None:
    """Test getting a property of the device."""
    metering_plug = local_gateway.devices[ELEMENT_ID]
    assert metering_plug.get_property("binary_switch") == list(metering_plug.binary_switch_property.values())
    assert metering_plug.get_property("consumption") == list(metering_plug.consumption_property.values())


def test_getting_state(local_gateway: HomeControl) -> None:
    """Test getting the online state of the device."""
    metering_plug = local_gateway.devices[ELEMENT_ID]
    assert metering_plug.is_online()
