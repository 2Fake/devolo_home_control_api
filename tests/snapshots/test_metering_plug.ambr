# serializer version: 1
# name: test_getting_devices
  list([
    Zwave(
      assistants_enabled=1,
      binary_sensor_property=dict({
        'devolo.BinarySensor:hdm:ZWave:CBC56091/2#Alarm(0)': BinarySensorProperty(
          device_uid='hdm:ZWave:CBC56091/2',
          element_uid='devolo.BinarySensor:hdm:ZWave:CBC56091/2#Alarm(0)',
          last_activity=datetime.datetime(1970, 1, 1, 1, 0, tzinfo=tzfile('/usr/share/zoneinfo/Europe/Berlin')),
          sensor_type='alarm',
          state=False,
          sub_type='overload',
        ),
        'devolo.WarningBinaryFI:hdm:ZWave:CBC56091/2': BinarySensorProperty(
          device_uid='hdm:ZWave:CBC56091/2',
          element_uid='devolo.WarningBinaryFI:hdm:ZWave:CBC56091/2',
          last_activity=datetime.datetime(1970, 1, 1, 1, 0, tzinfo=tzfile('/usr/share/zoneinfo/Europe/Berlin')),
          sensor_type='warning',
          state=False,
          sub_type='overload',
        ),
      }),
      binary_switch_property=dict({
        'devolo.BinarySwitch:hdm:ZWave:CBC56091/2': BinarySwitchProperty(
          device_uid='hdm:ZWave:CBC56091/2',
          element_uid='devolo.BinarySwitch:hdm:ZWave:CBC56091/2',
          enabled=True,
          last_activity=datetime.datetime(1970, 1, 1, 1, 0, tzinfo=tzfile('/usr/share/zoneinfo/Europe/Berlin')),
          state=False,
        ),
      }),
      brand='Everspring',
      consumption_property=dict({
        'devolo.Meter:hdm:ZWave:CBC56091/2': ConsumptionProperty(
          current=10,
          current_unit='W',
          device_uid='hdm:ZWave:CBC56091/2',
          element_uid='devolo.Meter:hdm:ZWave:CBC56091/2',
          last_activity=datetime.datetime(1970, 1, 1, 1, 0, tzinfo=tzfile('/usr/share/zoneinfo/Europe/Berlin')),
          total=28.38,
          total_since=datetime.datetime(2017, 5, 30, 8, 11, 4, 998000, tzinfo=tzfile('/usr/share/zoneinfo/Europe/Berlin')),
          total_unit='kWh',
        ),
      }),
      description=None,
      device_model_uid='devolo.model.Wall:Plug:Switch:and:Meter',
      device_type='Door Lock Keypad',
      element_uids=list([
        'devolo.Meter:hdm:ZWave:CBC56091/2',
        'devolo.BinarySwitch:hdm:ZWave:CBC56091/2',
        'devolo.BinarySensor:hdm:ZWave:CBC56091/2#Alarm(0)',
        'devolo.VoltageMultiLevelSensor:hdm:ZWave:CBC56091/2',
        'devolo.WarningBinaryFI:hdm:ZWave:CBC56091/2',
      ]),
      generic_device_class='Entry Control',
      href='https://mydevolo.com/v1/zwave/products/0x0060/0x0001/0x0002',
      identifier='SP814-US',
      is_own=True,
      is_securely_included=False,
      is_zwave_plus=True,
      manufacturer_id='0x0060',
      multi_level_sensor_property=dict({
        'devolo.VoltageMultiLevelSensor:hdm:ZWave:CBC56091/2': MultiLevelSensorProperty(
          device_uid='hdm:ZWave:CBC56091/2',
          element_uid='devolo.VoltageMultiLevelSensor:hdm:ZWave:CBC56091/2',
          last_activity=datetime.datetime(1970, 1, 1, 1, 0, tzinfo=tzfile('/usr/share/zoneinfo/Europe/Berlin')),
          sensor_type='voltage',
          sub_type='',
          unit='V',
          value=233.5,
        ),
      }),
      name='Everspring PIR Sensor SP814',
      pending_operations=False,
      product_id='0x0002',
      product_type_id='0x0001',
      secure_inclusion_code=0,
      setting_uids=list([
        'gds.hdm:ZWave:CBC56091/2',
        'cps.hdm:ZWave:CBC56091/2',
        'lis.hdm:ZWave:CBC56091/2',
        'ps.hdm:ZWave:CBC56091/2',
        'mas.hdm:ZWave:CBC56091/2',
      ]),
      settings_property=dict({
        'flash_mode': SettingsProperty(
          element_uid='mas.hdm:ZWave:CBC56091/2',
          last_activity=datetime.datetime(1970, 1, 1, 1, 0, tzinfo=tzfile('/usr/share/zoneinfo/Europe/Berlin')),
          value=3,
        ),
        'general_device_settings': SettingsProperty(
          element_uid='gds.hdm:ZWave:CBC56091/2',
          events_enabled=True,
          icon='light-bulb',
          last_activity=datetime.datetime(1970, 1, 1, 1, 0, tzinfo=tzfile('/usr/share/zoneinfo/Europe/Berlin')),
          name='Light Bulb',
          zone='Office',
          zone_id='hz_1',
        ),
        'led': SettingsProperty(
          element_uid='lis.hdm:ZWave:CBC56091/2',
          last_activity=datetime.datetime(1970, 1, 1, 1, 0, tzinfo=tzfile('/usr/share/zoneinfo/Europe/Berlin')),
          led_setting=True,
        ),
        'param_changed': SettingsProperty(
          element_uid='cps.hdm:ZWave:CBC56091/2',
          last_activity=datetime.datetime(1970, 1, 1, 1, 0, tzinfo=tzfile('/usr/share/zoneinfo/Europe/Berlin')),
          param_changed=False,
        ),
        'protection': SettingsProperty(
          element_uid='ps.hdm:ZWave:CBC56091/2',
          last_activity=datetime.datetime(1970, 1, 1, 1, 0, tzinfo=tzfile('/usr/share/zoneinfo/Europe/Berlin')),
          local_switching=True,
          remote_switching=True,
        ),
      }),
      specific_device_class='Secure Keypad Door Lock',
      status=2,
      uid='hdm:ZWave:CBC56091/2',
      zwave_version='6.51.07',
    ),
  ])
# ---
